### PR TITLE
Release long lived JSI objects ASAP

### PR DESCRIPTION
**Pull request by vmoroz**
### Type of Change
- Bug fix (non-breaking change which fixes an issue)

### Why
While investigating memory leaks in RNW code I saw that we almost never delete objects from the JSI long lived object collection.

### What
In this PR we release JSI long-lived objects as soon as we finish using them.
After this change, the list of long-lived objects is regularly cleaned.

### Backports:
- [0.72] #12388
- [0.73] #12387

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/12385)

### CODE REVIEW

**Review by JunielKatarn — APPROVED** [review]

**Comment by jonthysell** [comment]
@vmoroz It looks like the fabric e2e test app, using hermes, is crashing reliably in the PR checks. Is this a regression from your change?

File: vnext/Microsoft.ReactNative/TurboModulesProvider.cpp [hunk 1]
```diff
@@ -384,10 +436,6 @@ std::vector<std::string> TurboModulesProvider::getEagerInitModuleNames() noexcep
   if (it != m_moduleProviders.end()) {
     eagerModules.push_back("UIManager");
   }
-  it = m_moduleProviders.find("FabricUIManagerBinding");
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/TurboModulesProvider.cpp:387]: Why is this removed here and not in the backported changes to 73/72?

Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/TurboModulesProvider.cpp:387]: This looks like a merge conflict.  This code shouldn't be removed.

Comment (vmoroz) [reply vnext/Microsoft.ReactNative/TurboModulesProvider.cpp:387]: It is a bad merge. Thank you for catching it!

Comment (vmoroz) [reply vnext/Microsoft.ReactNative/TurboModulesProvider.cpp]: The lines are added back

**Comment by vmoroz** [comment]
> @vmoroz It looks like the fabric e2e test app, using hermes, is crashing reliably in the PR checks. Is this a regression from your change?

I am not sure. Is it in this PR, or it is part of the latest code? 
I.e. could you provide more info?

File: vnext/Microsoft.ReactNative/JSDispatcherWriter.cpp [hunk 2]
```diff
@@ -49,17 +55,18 @@ void JSDispatcherWriter::WithResultArgs(
...
-        args.reserve(dynValue.size());
-        auto &runtime = jsiRuntimeHolder->Runtime();
-        for (auto const &item : dynValue) {
-          args.emplace_back(facebook::jsi::valueFromDynamic(runtime, item));
-        }
-        handler(runtime, args.data(), args.size());
-      }
-    });
+    m_jsDispatcher.Post(
+        [handler, dynValue = std::move(dynValue), weakJsiRuntimeHolder = m_jsiRuntimeHolder, self = get_strong()]() {
```
Comment (JunielKatarn) [inline vnext/Microsoft.ReactNative/JSDispatcherWriter.cpp:59]: What's the purpose of the `self` reference if it's not directly used in the lambda?

**Review by JunielKatarn — APPROVED** [review]

File: vnext/Microsoft.ReactNative/JSDispatcherWriter.cpp (hunk 2 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/JSDispatcherWriter.cpp:59]: We must keep the JSDispatchWriter alive until we execute the code.
